File: server-Go/tetris/tetris.go
package tetris

import (
	"encoding/json"
	"log"
	"net/http"
	"strings"

	u "betazone/_utils"
)

type tetris struct {
	HighScoreFile string `json:"hsJSONFile"`
	HighScore     []struct {
		Player string `json:"player"`
		Score  int    `json:"score"`
	}
}

// Router ...
func (t *tetris) Router(w http.ResponseWriter, r *http.Request) {
	option := strings.ToLower(strings.Split(r.URL.Path, "/")[2])
	switch option {
	case "hs":
		if r.Method == "POST" {
			t.postHighScore(w, r)
			return
		}
		if r.Method == "GET" {
			t.getHighScore(w, r)
			return
		}
	}
	u.BadRequest(w, r)
}

func (t *tetris) getHighScore(w http.ResponseWriter, r *http.Request) {
	//u.LoadJSONFile(t.HighScoreFile, &t.HighScore)
	u.SendJSONToClient(w, &t.HighScore)
}

func (t *tetris) postHighScore(w http.ResponseWriter, r *http.Request) {
	decoder := json.NewDecoder(r.Body)
	err := decoder.Decode(&t.HighScore)
	if err != nil {
		log.Printf("Error saving HighScore = %s", err.Error())
		return
	}
	u.WriteJSONtoFile(t.HighScoreFile, t.HighScore)
}

func NewTetris() tetris {
	t := tetris{ //*new(tetris)
		HighScoreFile: "./tetris/highScore.json",
		HighScore:     nil,
	}
	u.LoadJSONFile(t.HighScoreFile, &t.HighScore)
	return t
}

/*
curl --header "Content-Type: application/json" --request POST --data '[
  {
    "player": "AAA",
    "score": 1000
  },
  {
    "player": "BBB",
    "score": 800
  },
  {
    "player": "CCC",
    "score": 600
  },
  {
    "player": "DDD",
    "score": 400
  },
  {
    "player": "EEE",
    "score": 200
  }
]
' http://localhost:3550/tetris/hs
*/

File: server-Node/tetris/tetris.js
/* */

import express from "express";
const tetris = express.Router();

import bodyParser from "body-parser";

import fs from 'fs';
import { dirname } from 'path';
import { fileURLToPath } from 'url';
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const filePath = __dirname + '/highScore.json';

tetris.use(bodyParser.json()); // to support JSON-encoded bodies
tetris.use(bodyParser.urlencoded({ // to support URL-encoded bodies
  extended: true
}));

tetris.use(express.json());

tetris.get("/hs", function (req, res) {
  loadJSONfile(filePath, function (highScore) {
    res
      .status(200)
      .send(highScore);
  });
});

tetris.post("/hs", function (req, res) {
  let score = req.body;
  console.log('==>', req.body);
  for (let i in score) {
    if (score[i].player.length > 2) {
      score[i].player = score[i].player.slice(0, 3);
    }
  }
  writeJSONtoFile(filePath, score, function () {
    res.end();
  });
  res.end();
});

export { tetris };

function loadJSONfile(filePath, callback) {
  fs.readFile(filePath, 'utf8', (err, data) => {
    if (err) {
      console.log(err);
    } else {
      callback(JSON.parse(data));
    }
  });
}

function writeJSONtoFile(filePath, dataSet, callback) {
  const dataJson = JSON.stringify(dataSet);
  console.log(dataJson);
  fs.writeFile(filePath, dataJson, 'utf8', callback);
}

/*
curl --header "Content-Type: application/json" --request POST --data '[
  {
    "player": "AAA",
    "score": 1000
  },
  {
    "player": "BBB",
    "score": 800
  },
  {
    "player": "CCC",
    "score": 600
  },
  {
    "player": "DDD",
    "score": 400
  },
  {
    "player": "EEE",
    "score": 200
  }
]
' http://localhost:3000/tetris/hs
*/
